packages/transform/src/lib/TransformMiddleware.ts

```diff
@@ -0,0 +1,485 @@
+    const { key, path, value } = payload;
+    const { before } = this.context;
+
+    payload.value = (await before(value as unknown as BeforeValue, key, path)) as ReturnValue;
+    await this.provider[Method.Push](payload);
+    await this.updateMetadataPath(key, path);
+
+    return payload;
+  }
+
+  @PreProvider()
+  public async [Method.Set]<StoredValue = AfterValue>(payload: Payload.Set<StoredValue>): Promise<Payload.Set<StoredValue>> {
+    const { key, path, value } = payload;
+    const { before } = this.context;
+
+    payload.value = (await before(value as unknown as BeforeValue, key, path ?? null)) as StoredValue;
+    await this.provider[Method.Set](payload);
+    if (path) {
+      await this.updateMetadataPath(key, path);
+      return payload;
+    }
+
+    const { data } = await this.provider[Method.Get]({ method: Method.Get, errors: [], key, path: [] });
+
+    await this.updateMetadataPath(key, this.objectPathKeys(data));
+
+    return payload;
+  }
+
+  @PreProvider()
+  public async [Method.SetMany](payload: Payload.SetMany): Promise<Payload.SetMany> {
+    const { entries } = payload;
+    const { before } = this.context;
+
+    payload.entries = entries.map(({ key, path, value }) => ({ key, path, value: before(value as BeforeValue, key, path ?? null) }));
+    await this.provider[Method.SetMany](payload);
+
+    entries.forEach(async ({ key, path }) => {
+      const { data } = await this.provider[Method.Get]({ method: Method.Get, errors: [], key, path });
+
+      await this.updateMetadataPath(key, this.objectPathKeys(data));
+    });
+
+    return payload;
+  }
+
+  @PreProvider()
+  public async [Method.Update]<StoredValue = AfterValue, ReturnValue = BeforeValue>(
+    payload: Payload.Update<StoredValue, ReturnValue>
+  ): Promise<Payload.Update<StoredValue, ReturnValue>> {
+    const { key, hook } = payload;
+    const { before } = this.context;
+    const getBefore = await this.provider[Method.Get]({ method: Method.Get, errors: [], key, path: [] });
+
+    payload.hook = (value: StoredValue) => before(hook!(value, key) as BeforeValue, key, null) as Awaitable<ReturnValue>;
+
+    await this.provider[Method.Update](payload as unknown as Payload.Update<AfterValue, ReturnValue>);
+
+    const getAfter = await this.provider[Method.Get]({ method: Method.Get, errors: [], key, path: [] });
+
+    if (this.getChangedKeys(getBefore.data, getAfter.data).length! > 0) {
+      await this.updateMetadataPath(key, this.getChangedKeys(getBefore.data, getAfter.data));
+    }
+
+    return payload;
+  }
+
+  @PostProvider()
+  public async [Method.Dec](payload: Payload.Dec): Promise<Payload.Dec> {
+    const { after, before } = this.context;
+    const { key } = payload;
+    const getBefore = await this.provider[Method.Get]({ method: Method.Get, errors: [], key, path: [] });
+    const beforeValue = after(getBefore.data!, key, null);
```

Comment:
Why would data be definitely defined here? I would recommend using isPayloadWithData and destructuring data in order to check if data is defined